File: packages/dart/oauth/lib/src/token_storage.dart
import 'dart:convert';
import 'dart:io';
import 'package:path/path.dart' as path;

/// File-based token storage for OAuth tokens.
/// Stores tokens in ~/.oauth.do/token
class TokenStorage {
  static const String _defaultTokenDir = '.oauth.do';
  static const String _defaultTokenFile = 'token';

  final String tokenPath;

  /// Creates a TokenStorage with default path (~/.oauth.do/token).
  TokenStorage()
      : tokenPath = path.join(
          Platform.environment['HOME'] ?? Platform.environment['USERPROFILE'] ?? '.',
          _defaultTokenDir,
          _defaultTokenFile,
        );

  /// Creates a TokenStorage with a custom path.
  TokenStorage.custom(this.tokenPath);

  /// Saves token data to disk.
  Future<void> save(TokenData data) async {
    final file = File(tokenPath);
    await file.parent.create(recursive: true);

    final json = jsonEncode(data.toJson());
    await file.writeAsString(json);

    // Set restrictive permissions on Unix systems
    if (!Platform.isWindows) {
      await Process.run('chmod', ['600', tokenPath]);
    }
  }

  /// Loads token data from disk.
  Future<TokenData?> load() async {
    final file = File(tokenPath);
    if (!await file.exists()) {
      return null;
    }

    try {
      final content = await file.readAsString();
      final json = jsonDecode(content) as Map<String, dynamic>;
      return TokenData.fromJson(json);
    } catch (_) {
      return null;
    }
  }

  /// Deletes stored token data.
  Future<bool> delete() async {
    final file = File(tokenPath);
    try {
      if (await file.exists()) {
        await file.delete();
        return true;
      }
      return false;
    } catch (_) {
      return false;
    }
  }

  /// Checks if a valid (non-expired) token exists.
  Future<bool> hasValidToken() async {
    final data = await load();
    return data != null && !data.isExpired;
  }

  /// Gets the token file path.
  String getTokenPath() => tokenPath;
}

/// Token data structure for serialization.
class TokenData {
  final String accessToken;
  final String? refreshToken;
  final String tokenType;
  final int expiresAt;

  TokenData({
    required this.accessToken,
    this.refreshToken,
    this.tokenType = 'Bearer',
    required this.expiresAt,
  });

  /// Creates TokenData from JSON.
  factory TokenData.fromJson(Map<String, dynamic> json) {
    return TokenData(
      accessToken: json['accessToken'] as String,
      refreshToken: json['refreshToken'] as String?,
      tokenType: json['tokenType'] as String? ?? 'Bearer',
      expiresAt: json['expiresAt'] as int,
    );
  }

  /// Converts TokenData to JSON.
  Map<String, dynamic> toJson() {
    return {
      'accessToken': accessToken,
      'refreshToken': refreshToken,
      'tokenType': tokenType,
      'expiresAt': expiresAt,
    };
  }

  /// Checks if the token is expired.
  bool get isExpired {
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    return now >= expiresAt;
  }
}

File: packages/php/oauth/src/TokenStorage.php
<?php

declare(strict_types=1);

namespace DotDo\OAuth;

/**
 * File-based token storage.
 */
class TokenStorage
{
    private string $tokenDir;
    private string $tokenFile;

    public function __construct(?string $tokenDir = null)
    {
        $this->tokenDir = $tokenDir ?? $this->getDefaultTokenDir();
        $this->tokenFile = $this->tokenDir . '/token';
    }

    private function getDefaultTokenDir(): string
    {
        $home = getenv('HOME') ?: getenv('USERPROFILE');
        return $home . '/.oauth.do';
    }

    /**
     * Saves tokens to file storage.
     *
     * @param array $tokens Token data
     */
    public function saveTokens(array $tokens): void
    {
        if (!is_dir($this->tokenDir)) {
            mkdir($this->tokenDir, 0700, true);
        }

        file_put_contents(
            $this->tokenFile,
            json_encode($tokens, JSON_PRETTY_PRINT),
            LOCK_EX
        );
        chmod($this->tokenFile, 0600);
    }

    /**
     * Loads tokens from file storage.
     *
     * @return array|null Token data or null if not found
     */
    public function loadTokens(): ?array
    {
        if (!file_exists($this->tokenFile)) {
            return null;
        }

        $content = file_get_contents($this->tokenFile);
        if ($content === false) {
            return null;
        }

        return json_decode($content, true);
    }

    /**
     * Gets the access token from storage.
     *
     * @return string|null Access token or null if not found
     */
    public function getAccessToken(): ?string
    {
        $tokens = $this->loadTokens();
        return $tokens['access_token'] ?? null;
    }

    /**
     * Gets the refresh token from storage.
     *
     * @return string|null Refresh token or null if not found
     */
    public function getRefreshToken(): ?string
    {
        $tokens = $this->loadTokens();
        return $tokens['refresh_token'] ?? null;
    }

    /**
     * Deletes stored tokens.
     */
    public function deleteTokens(): void
    {
        if (file_exists($this->tokenFile)) {
            unlink($this->tokenFile);
        }
    }

    /**
     * Checks if tokens exist.
     */
    public function hasTokens(): bool
    {
        return file_exists($this->tokenFile);
    }
}
